op-chain-ops/cmd/celo-migrate/main.go

```diff
@@ -515,115 +515,34 @@ func runDBCheck(opts dbCheckOptions) (err error) {
+	// Check the range covering the ancient to non-ancient transition first, since this is where we previously encountered continuity issues.
+	err = checkRange(ancientDB, nonAncientDB, ancients-1, ancients+1)
 	if err != nil {
-		if opts.failFast {
-			return fmt.Errorf("failed to load first non-ancient block: %w", err)
-		}
-		// We don't need to add the error to errResult here because it will be added below when we call checkContinuity on non-ancients
-	} else {
-		if _, err := firstNonAncientRange.CheckContinuity(lastAncient, 1); err != nil {
-			err = fmt.Errorf("failed continuity check between ancients and non-ancients: %w", err)
-			if opts.failFast {
-				return err
-			}
-			errResult = multierror.Append(errResult, err)
-		}
+		return fmt.Errorf("failed to check range covering ancient to non-ancient transition (blocks %d-%d): %w", ancients-1, ancients+1, err)
 	}
-
-	ctx, cancel := context.WithCancel(context.Background())
-	defer cancel()
-	g, ctx := errgroup.WithContext(ctx)
-	g.SetLimit(8)
-
-	var mu sync.Mutex
-
-	checkRange := func(start, count uint64, loadRangeFunc func(uint64, uint64) (*RLPBlockRange, error)) {
-		// If we are not at genesis or the first non-ancient block, include the last block of
-		// the previous range so we can check for continuity between ranges.
-		if start != 0 && start != lastAncientNumber+1 {
-			start = start - 1
-			count = count + 1
-		}
+	// Now check the whole chain
+	g := new(errgroup.Group)
+	g.SetLimit(runtime.NumCPU() * 2)
+	lastBlockNumber := *rawdb.ReadHeaderNumber(nonAncientDB, rawdb.ReadHeadHeaderHash(nonAncientDB))
+	for i := uint64(0); i <= lastBlockNumber; i += opts.batchSize {
+		start := i
 		g.Go(func() error {
-			select {
-			case <-ctx.Done():
-				return ctx.Err()
-			default:
-				blockRange, err := loadRangeFunc(start, count)
-				if err != nil {
-					err = fmt.Errorf("failed to load block range: %w", err)
-					if opts.failFast {
-						return err
-					}
-					log.Error(err.Error())
-					mu.Lock()
-					errResult = multierror.Append(errResult, err)
-					mu.Unlock()
-					return nil
-				}
-				if _, err := blockRange.CheckContinuity(nil, count); err != nil {
-					err = fmt.Errorf("failed continuity check: %w", err)
-					if opts.failFast {
-						return err
-					}
-					log.Error(err.Error())
-					mu.Lock()
-					errResult = multierror.Append(errResult, err)
-					mu.Unlock()
-					return nil
-				}
-				log.Info("Successfully checked block range continuity", "start", start, "end", start+count-1, "count", count)
-				return nil
+			end := min(start+opts.batchSize-1, lastBlockNumber)
+			if start > 0 {
+				start -= 1 // Ensure we check continuity across batch boundaries
+			}
+			log.Info("Checking data continuity for block range", "start", start, "end", end, "count", end-start+1)
+			err := checkRange(ancientDB, nonAncientDB, start, end)
+			if err != nil {
+				return fmt.Errorf("failed to check range (%d-%d): %w", start, end, err)
 			}
+			return nil
```

Comment:
Don't you need to have a call to g.Wait() before returning?

Reply:
Tested this and it looks like errors thrown from within the go routine do not cause the script to fail for this reason

Reply:
Yes we do need that, thanks!